### ISSUE TITLE
Should da ufuncs be lazy or immediate on numpy arrays?

### ISSUE THREAD

**Issue by mrocklin** [dask/dask#2103]
Currently functions like `da.sin` on a numpy array compute immediately.  However all other cases of `atop` wrap the numpy array into a single-chunk dask.array and compute lazily.  Should we change the immediate behavior of dask.array ufuncs to be lazy?

cc @shoyer @jhamman

**Comment by pelson** [comment]
As per the sentiment in https://github.com/dask/dask/pull/2301#issuecomment-310097823:

> Dask's convention so far is to call the equivalent numpy method in these cases.

Personally, I'd expect to always have a lazy dask thing if I've called dask functions/methods. If I wanted immediate, I'd call numpy directly...

---

Is that an overly simplistic view?

**Comment by jakirkham** [comment]
I agree with Phil that ufuncs should be lazy. Naively that is what I would have expected.

> Is that an overly simplistic view?

Well there are some funny cases, but they are a bit off topic. For instance, handling conditional statements are not lazy (and probably shouldn't be IMHO). Please see the (contrived) example below for context.

```python
In [1]: import dask.array as da

In [2]: d = da.random.random((10, 12), chunks=(5, 6))

In [3]: if ((0 < d) | (d < 1)).any():
   ...:     print("Things are fine.")
   ...: else:
   ...:     print("Uh oh.")
   ...:     
Things are fine.

In [4]: ((0 < d) | (d < 1)).any().__nonzero__()
Out[4]: True
```

**Comment by GenevieveBuckley** [comment]
> Probably have these `ufunc`s call `da.asarray` on any NumPy arrays they are given?

Is there a central place in the code where we can do this for all `ufunc`, or would this need to be done for each one individually?

I'm wondering if this would make a good first or second issue.
